aspnetcore/security/authentication/mfa.md

```diff
@@ -0,0 +1,560 @@
+
+	services.AddSingleton<IEmailSender, EmailSender>();
+	services.AddScoped<IUserClaimsPrincipalFactory<IdentityUser>, 
+		AdditionalUserClaimsPrincipalFactory>();
+
+	services.AddAuthorization(options =>
+	{
+		options.AddPolicy("TwoFactorEnabled",
+			x => x.RequireClaim("TwoFactorEnabled", "true" )
+		) ;
+	});
+
+	services.AddRazorPages();
+}
+```
+
+The AdditionalUserClaimsPrincipalFactory adds the **TwoFactorEnabled** claim to the user claims after a successful login. This is only added after a login. The value is read from the database. This is added here because the user should only access the higher protected view, if the identity has logged in with MFA. If the database view was read from the database directly instead of using the claim, it would be possible to access the view without MFA directly after activating the MFA.
+
+```csharp
+using Microsoft.AspNetCore.Identity;
+using Microsoft.Extensions.Options;
+using System.Collections.Generic;
+using System.Security.Claims;
+using System.Threading.Tasks;
+
+namespace IdentityStandaloneMfa
+{
+    public class AdditionalUserClaimsPrincipalFactory : UserClaimsPrincipalFactory<IdentityUser, IdentityRole>
+    {
+        public AdditionalUserClaimsPrincipalFactory( 
+            UserManager<IdentityUser> userManager,
+            RoleManager<IdentityRole> roleManager, 
+            IOptions<IdentityOptions> optionsAccessor) 
+            : base(userManager, roleManager, optionsAccessor)
+        {
+        }
+
+        public async override Task<ClaimsPrincipal> CreateAsync(IdentityUser user)
+        {
+            var principal = await base.CreateAsync(user);
+            var identity = (ClaimsIdentity)principal.Identity;
+
+            var claims = new List<Claim>();
+
+            if (user.TwoFactorEnabled)
+            {
+                claims.Add(new Claim("TwoFactorEnabled", "true"));
+            }
+            else
+            {
+                claims.Add(new Claim("TwoFactorEnabled", "false")); ;
+            }
+
+            identity.AddClaims(claims);
+            return principal;
+        }
+    }
+}
+```
+
+Because we changed the Identity service setup in the Startup class, the layouts of the Identity need to be updated. Scaffold the Identity pages into the application. Define the layout in the Identity/Account/Manage/_Layout.cshtml file.
+
+```cshtml
+@{
+    Layout = "/Pages/Shared/_Layout.cshtml";
+}
+```
+
+Also add the _layout for all the manage pages from the Identity Pages.
+
+```cshtml
+@{
+    Layout = "_Layout.cshtml";
+}
+```
+
+### Validate the MFA requirement in the admin page
+
+The admin Razor Page validates that the user has logged in using MFA. In the OnGet method, the identity is used to access the user claims. The **TwoFactorEnabled** claim is checked for the value true. If the identity is missing this claim, or it is false, the page will redirect to the Enable MFA page. This is possible because the user has logged in already, but without MFA.
```

Comment:
```suggestion
The administration Razor Page validates that the user has logged in using MFA. In the `OnGet` method, the identity is used to access the user claims. The `TwoFactorEnabled` claim is checked for the value `true`. If the identity is missing this claim or is false, the page redirects to the Enable MFA page. This is possible because the user has logged in already, but without MFA.
```